src/ui/suspense/nodeCaches.ts

```diff
@@ -173,27 +160,157 @@ export const nodeDataCache: Cache<
+  NodeInfo[] | null
+> = createCache({
+  config: { immutable: true },
+  debugLabel: "AncestorNodesData",
+  getKey: ([replayClient, pauseId, nodeId]) => `${pauseId}:${nodeId}`,
+  load: async ([replayClient, pauseId, nodeId]): Promise<NodeInfo[] | null> => {
+    if (!pauseId || typeof nodeId !== "string") {
+      return null;
+    }
+
+    const nodes: NodeInfo[] = [];
+
+    let ancestorId: string | null = nodeId;
+
+    while (ancestorId) {
+      const node = await processedNodeDataCache.readAsync(replayClient, pauseId, ancestorId);
+      ancestorId = node?.parentNodeId ?? null;
+      if (node) {
+        nodes.unshift(node);
+      }
+    }
+
+    return nodes;
+  },
+});
+
+export const reIsNotWhiteSpace = /[^\s]/;
+
+export const canRenderNodeInfo = (childNode: NodeInfo | null): childNode is NodeInfo => {
+  if (!childNode) {
+    return false;
+  }
+
+  const canRender = !!(
+    childNode.type !== NodeConstants.TEXT_NODE || reIsNotWhiteSpace.exec(childNode.value!)
+  );
+  return canRender;
+};
+
+export const getCurrentRenderableChildNodeIds = (
+  replayClient: ReplayClientInterface,
+  pauseId: string,
+  parentNode: NodeInfo,
+  renderableChildNodes: NodeInfo[] | null | undefined
+) => {
+  let childNodeIds: string[] = [];
+
+  if (renderableChildNodes) {
+    // We have the complete list of all _renderable_ child nodes.
+    childNodeIds = renderableChildNodes.map(node => node.id);
+  } else {
+    // Only try to render children that are already in cache.
+    // This should only happen when the user selects a deeply nested node,
+    // in which case we have _some_ children available but may not have all.
+    // The renderable cache request should resolve shortly and then
+    // we'll use that list instead.
+    childNodeIds = parentNode.children.filter(childNodeId => {
+      const maybeNode = processedNodeDataCache.getValueIfCached(replayClient, pauseId, childNodeId);
+      return !!maybeNode && canRenderNodeInfo(maybeNode);
+    });
+  }
+
+  return childNodeIds;
+};
+
+// TODO [FE-1846???] The backend returns _all_ child nodes in the preview,
+// and we have no way of knowing which ones are renderable..
+// We've asked for a way to filter out the ones that aren't renderable on the server.
+// Until then, we'll filter them out here.
+// ref: https://linear.app/replay/issue/FE-1846/investigate-honeycomb-telemetry-that-can-help-us-see-which-getboxmodel#comment-0fb1c7e6
+// TODO [BAC-3918]Similarly, this would _really_ benefit from
+// a way to bulk-fetch object previews
+export const renderableChildNodesCache: Cache<
+  [replayClient: ReplayClientInterface, pauseId: PauseId, nodeId: string],
+  NodeInfo[] | null
+> = createCache({
+  config: { immutable: true },
+  debugLabel: "ChildNodesData",
+  getKey: ([replayClient, pauseId, nodeId]) => `${pauseId}:${nodeId}`,
+  load: async ([replayClient, pauseId, nodeId]): Promise<NodeInfo[] | null> => {
+    if (!pauseId || typeof nodeId !== "string") {
```

Comment:
This is unexpected, given the way the types are defined for this cache:
```js
  [replayClient: ReplayClientInterface, pauseId: PauseId, nodeId: string],
```
Which is wrong?

Reply:
The argument types, so I should presumably update those (which will have ripple effects).

This is an example of why I was asking for a `skip` option for the `useImperativeCacheValue` hook.  There _are_ cases where we will not have a valid pause ID or node ID, but right now we can't skip calling into the cache.

Reply:
Seems like updating the types is appropriate then I guess.